### ISSUE TITLE
[css-scroll-snap-1] Avoid page scrolling skipping past snappable items

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#10914]
This originally came up from @johannesodland as a potential issue with scroll-buttons in https://github.com/w3c/csswg-drafts/issues/10722#issuecomment-2353648101 however it's a pre-existing generic issue with scroll-snap, so filing it separately.

When a user pages down, usually by pressing the page-down key, they expect to scroll no further than one page if that is a valid scroll location. However, some browsers can select the next snap point as the target resulting in a scroll of more than a scrollport in length.

E.g. if you have a scrollport height of 1000px, most browsers will try to scroll by 850px. If however the nearest snappable element has a snap alignment that is further, then it can be selected, resulting in scrolls even greater than 1000px and an experience where content has been skipped over - even when that content itself defined a valid snap area. We should fix this so that it's easy to make an experience where 

Codepen demo using page-down key: https://codepen.io/flackr/pen/abgeOKY
Original demo from @johannesodland using `scrollTo` api: https://codepen.io/johannesodland/pen/WNqmoYy

**Comment by flackr** [comment]
Naively for built-in interactions, the UA could do something like select a snap area which scrolls no further than 100% of the scrollport if possible, though I wonder if there's some way we could make this work for programmatic scrolls as well as in the scrollTo example?

E.g. we could say that for any scrolls less than or equal to 1 scrollport, we try to select a snap area which keeps the scroll less than that distance. I say "try" because developers can position snap areas arbitrarily far apart and we should make some progress in the requested scrolling direction.

**Comment by flackr** [comment]
From @dholbert:
> FWIW there is a scrollByPages API in Firefox, non-standard:  https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollByPages

**Comment by dholbert** [comment]
> From @dholbert:
> 
> > FWIW there is a scrollByPages API in Firefox, non-standard:  https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollByPages

Indeed; I'm pretty sure this is how we implement the `Space/PageDown` and `PageUp` scrolling actions under-the-hood (as calls to `scrollByPages(1)` and `scrollByPages(-1)`, from Firefox's frontend JS).  But the JS API is also exposed to the web, too, for historical reasons I guess.

The [`scrollByPages()` implementation is here](https://searchfox.org/mozilla-central/rev/72d959d715d3f832328057600811c09ee5195ae1/dom/base/nsGlobalWindowInner.cpp#3909-3926) if anyone's curious to see it.  That function calls a generic C++ `ScrollBy()` function that's shared by a bunch of programmatic scrolling APIs, and the most relevant piece there for `scrollByPages` is the  [`GetPageScrollAmount()`](https://searchfox.org/mozilla-central/rev/72d959d715d3f832328057600811c09ee5195ae1/layout/generic/ScrollContainerFrame.cpp#5161-5192) function, which defines what it means to scroll by a page.

**Comment by argyleink** [comment]
could we make snapping more conservative and choose the nearer of the points instead of the further?

**Comment by flackr** [comment]
> The [`scrollByPages()` implementation is here](https://searchfox.org/mozilla-central/rev/72d959d715d3f832328057600811c09ee5195ae1/dom/base/nsGlobalWindowInner.cpp#3909-3926) if anyone's curious to see it. That function calls a generic C++ `ScrollBy()` function that's shared by a bunch of programmatic scrolling APIs, and the most relevant piece there for `scrollByPages` is the [`GetPageScrollAmount()`](https://searchfox.org/mozilla-central/rev/72d959d715d3f832328057600811c09ee5195ae1/layout/generic/ScrollContainerFrame.cpp#5161-5192) function, which defines what it means to scroll by a page.

In this case we want to be a bit smarter about what it means to scroll by a page - in particular, avoiding scrolling further than a full page as a result of snapping to avoid skipping content.

Currently, the algorithm we have in chrome to do this does so by [having three constraints](https://source.chromium.org/chromium/chromium/src/+/main:cc/input/snap_selection_strategy.cc;l=55;drc=c8dc70b538f1bb0862f1be58237d6e945ee81819) - an optimal scroll distance which seems equivalent to `GetPageScrollAmount()` you pointed to above, as well as a minimum and maximum (a full visible screen of content). Then, this is used to find a snap target that is as close to the optimal distance but not further than the maximum (avoiding skipped content). If this is not possible (e.g. the next snap target is more than a page away), then it will skip further.

**Comment by flackr** [comment]
The simple approach which matches implementations would seem to be to add a granularity to [ScrollToOptions](https://www.w3.org/TR/cssom-view-1/#dictdef-scrolltooptions). If you passed a page granularity then it could interpret the `left` and `top` values as # of pages. Happy to bikeshed on names here and/or other options (e.g. it's common to scroll by lines).

However, given the algorithm for finding a snap area https://github.com/w3c/csswg-drafts/issues/10914#issuecomment-3497842094, if we naively use the given value, you may get different results if you `scrollBy({top: 1, granularity: 'page'}); scrollBy({top: 1, granularity: 'page'});` than if you `scrollBy({top: 2, granularity: 'page'});` as the first scroll in the first example may have chosen a snap area that is less than or more than the optimal page scroll distance resulting in a different range for the second scroll. I don't think this is necessarily a problem, just wanted to call it out.

If we wanted, we could expose a scroll by operation which took the 3 constraints we use in this algorithm for full generality and customizability, however I think it's fine to leave this as an internal detail of what it means to scroll by pages.

**Comment by shanthambi** [comment]
@flackr 

In the same setup, scroll-marker scrolls to the expected snap-aligned position, but scroll-button consistently scrolls only ~85% of the required distance, leaving the target slide partially visible.

Since both scroll-marker and scroll-button are part of the same CSS Carousel mechanism, they should resolve to the same scroll snap positions and produce identical final scroll offsets.

The current behavior makes scroll-button unreliable, even though the snap points themselves are correct (as demonstrated by scroll-marker).

From an authoring perspective, this inconsistency makes it impossible to rely on CSS carousel controls without JavaScript fallbacks, especially for predictable and accessible navigation.

I can also reproduce this using the demo below.
Tested in Chrome Version 143.0.7499.41.

This highlights an inconsistency between scroll-marker and scroll-button, which should both resolve to the same scroll snap positions.

https://codepen.io/shanthambi/pen/gbrNMXg
